[site] : site updates 5.0.0 iteration 1 (#2821)
exadel-inc/esl: src/modules/esl-event-listener/docs/1-overview.md

- `target` - definition of `EventTarget` element (or string `TraversingQuery` to find it, see details in [`ESLEventDesriptor`](#descriptors-esleventdesriptor-esleventdesriptorfn));
- `selector` - CSS selector to use built-in event delegation;
- `capture` - marker to use the capture phase of the DOM event life-cycle;
- `passive` - marker to use passive (non-blocking) subscription of the native event (if supported);
- `once` - marker to destroy the subscription after the first event catch.
- `group` - auxiliary property to group subscriptions. Does not affect the subscription behavior. Can be used for filtering and bulk operations.

All of the `ESLEventListener` instance fields are read-only; the subscription can't be changed once created.

The `ESLEventListener`, as a class, describes the subscription behavior and
contains static methods to create and manage subscriptions.

### <a name="descriptors-esleventdesriptor-esleventdesriptorfn">Descriptors (`ESLEventDesriptor`, `ESLEventDesriptorFn`)</a>

The event listener _Descriptor_ is an object to describe future subscriptions.
The ESL event listeners module has a few special details regarding such objects.

> A simple descriptor is an object that is passed to ESL event listener API to create a subscription.

Reviewer: ```suggestion
A simple descriptor is an object that is passed to ESLEventListener API to create a subscription.
```

  E.g.:
  ```typescript
  @listen({ event: 'click', selector: 'button' })
  onClick(e: DelegatedEvent<MouseEvent> /* instead of MouseEvent */) {
    const delegate = e.$delegate; //instaead of e.target && e.target.closest('button');
    ...
  }
  ```

  Supports `PropertyProvider` to declare the computed value as well.

- #### `condition` key

  <u>Type:</u> `bollean | PropertyProvider<boolean>`
  <u>Default Value:</u> `true`  
  <u>Description:</u> the function predicate or boolean flag to check if the subscription should be created. Resolves right before the subscription.

  Useful in combination with `@listen` decorator to declare subscriptions.

  ```typescript
    class MyEl extends ESLBaseElement {
        @attr() enabled = true;     
  
        @listen({event: 'click', condition: (that) => that.enabled})
        onClick(e) {}
  
        attributeChangedCallback(name, oldValue, newValue) {
          if (name === 'enabled') {
              ESLEventUtils.unsubscribe(this, this.onClick);
              ESLEventUtils.subscribe(this, this.onClick);
          }
        }
    }
  ```

- #### `capture` key

  <u>Type:</u> `boolean`  
  <u>Default Value:</u> `false`  

>   <u>Description:</u> marker to use capturing phase of the DOM event to handle.

Reviewer: ```suggestion
  <u>Description:</u> marker to use the capturing phase of the DOM event to handle.
```

  <u>Type:</u> `boolean`  
  <u>Default Value:</u> `false` for `ESLEventUtils.initDescriptor`, `true` for `@listen` decorator
  <u>Description:</u> marker to make an auto-subscribable descriptor. See [Automatic (collectable) descriptors](#automatic-collectable-descriptors).
- #### `inherit` key (for `ESLEventDesriptorExt` only)
  <u>Type:</u> `boolean`  
  <u>Description:</u> available in extended version of `ESLEventDesriptor` that is used in the descriptor declaration API.
  Allows to inherit `ESLEventDesriptor` data from the `ESLEventDesriptorFn` from the prototype chain.
  See [`initDescriptor`](#-esleventutilsinitdescriptor) usages example.

### <a name="automatic-collectable-descriptors">Automatic (collectable) descriptors</a>

Auto-collectable (or auto-subscribable) descriptors can be subscribed at once during the initialization of the `host` object.

To make an `ESLEventDesriptorFn` auto-collectable, the consumer should declare it with the `auto` marker using
`ESLEventUtils.initDescriptor` or `@listen` decorator.

⚠ `ESLEventUtils.initDescriptor` (or `@listen`) stores the auto-collectable descriptors in the internal collection on the `host`.

> The `ESLBaseElment` and the `ESLMixinElement` subscribes all auto-collectable descriptors in the `connectedCallback`.

Reviewer: ```suggestion
The `ESLBaseElment` and the `ESLMixinElement` subscribe all auto-collectable descriptors in the `connectedCallback`.
```
